### PR TITLE
[WEB-211] Account profile redesign

### PR DESCRIPTION
**Pull request by 314numberpi**

### CODE REVIEW

File: client/helpers/getDisplayRat.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const getDisplayRat = (user, ratState) => {
```
Comment (UncleClapton) [inline client/helpers/getDisplayRat.js]: Not sure how this got added back again. File should be removed.

File: client/components/UserDecalPanel.js [hunk 2]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
     return null
   }
 
+  _renderDecalCode = (decal, index) => (
+    <>
+      <div className="decal-code" key={`code${index}`}>
+        <button
+          key={`toggle${index}`}
+          aria-label="toggle-decal"
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: aria-label should be human readable explanatory text, as it is read by screen readers.
```suggestion
          aria-label="Show decal code"
```

File: client/components/UserDecalPanel.js [hunk 3]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
     return null
   }
 
+  _renderDecalCode = (decal, index) => (
+    <>
+      <div className="decal-code" key={`code${index}`}>
+        <button
+          key={`toggle${index}`}
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: unnecessary key being set here. only the root element returned in a map needs the key.

File: client/components/UserDecalPanel.js [hunk 4]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
     return null
   }
 
+  _renderDecalCode = (decal, index) => (
+    <>
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Avoid using fragments in maps when possible. This should probably be a wrapper `<div />` with an identifying className like `decal-row`

File: client/components/UserDecalPanel.js [hunk 5]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
...
+      <div className="decal-code" key={`code${index}`}>
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: You should never use `index` in a key attribute. 
Since we are mapping out a decal resource, we should use it's id

```suggestion
      <div className="decal-code" key={decal.id}>
```

File: client/components/UserDecalPanel.js [hunk 6]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
...
+
+  _renderDecalCodes = () => {
+    const { decals } = this.props
+    console.log(decals)
+
+    if (decals === []) {
+      return null
+    }
+
+    return (
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: no need for the parenthesis here. you can just do
```javascript
return decals.map(this._renderDecalCode)
```

File: client/components/UserDecalPanel.js [hunk 7]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
     return null
   }
 
+  _renderDecalCode = (decal, index) => (
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: render methods should be "public" (without the `_`), and placed directly above the main render method.

File: client/components/UserDecalPanel.js [hunk 8]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
...
+    <>
+      <div className="decal-code" key={`code${index}`}>
+        <button
+          key={`toggle${index}`}
+          aria-label="toggle-decal"
+          className="icon decal-toggle"
+          onClick={this._handleToggle}
+          // title={maxNicksReached ? 'You\'ve used all your nicknames' : 'Add new nickname'}
+          type="button">
+          <FontAwesomeIcon icon={this.state.formOpen[index] ? 'eye-slash' : 'eye'} fixedWidth />
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: again, avoid using index. Better to use the decal's id like we do with key attributes.

File: client/components/UserDecalPanel.js [hunk 9]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
...
+        </button>
+        {decal.attributes.code}
+      </div>
+      <div className="decal-claimed-at" key={`claimedAt${index}`}>{formatAsEliteDate(decal.attributes.claimedAt)}</div>
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: same suggestion on keys as above, however if we make the wrapper element a div then this can just be removed.

File: client/components/UserDecalPanel.js [hunk 10]
```diff
@@ -108,26 +142,47 @@ class UserDetailsPanel extends Component {
...
+      //   loadingText={checkingEligibility ? 'Checking decal eligibility...' : 'Redeeming decal codes...'}
+      //   manual
+      //   minRows={3}
+      //   noDataText={this._renderNoDataText()}
+      //   showPagination={false} />
+      <div className="panel user-decals">
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this._renderDecalCodes(decals))}
+          {(decals === []) ? (this._renderNoDataText()) : ('')}
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Same as above, never compare to `[]`

**Review by UncleClapton — CHANGES_REQUESTED** [review]

File: client/components/UserDecalPanel.js (hunk 4 above)
Comment (314numberpi) [reply client/components/UserDecalPanel.js]: The problem with that is it messes up the formatting, the grid requires all direct child elements to function correctly (sub-grids, when?), otherwise the date column would either have variable widths for each row, or a fixed-width, not based on the content.

File: client/components/UserDecalPanel.js (hunk 10 above)
Comment (314numberpi) [reply client/components/UserDecalPanel.js]: I don't know what you mean by this or why, or even how to do this otherwise?

Wait... length check perhaps...

Comment (UncleClapton) [reply client/components/UserDecalPanel.js]: wtf... the other one didn't get saved :/

File: client/components/UserDecalPanel.js [hunk 11]
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
...
+      </div>
+      <div className="decal-claimed-at" key={`claimedAt${index}`}>{formatAsEliteDate(decal.attributes.claimedAt)}</div>
+    </>
+  )
+
+  _renderDecalCodes = () => {
+    const { decals } = this.props
+    console.log(decals)
+
+    if (decals === []) {
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Never compare to `[]` as it will always return false. using `[]` always creates a new instance of an object, meaning regardless of contents the two will never be equal. Since decals is always an array, it's safe to do a truthy check on `decals.length`.
```suggestion
    if (!decals.length) {
```

**Review by UncleClapton — CHANGES_REQUESTED** [review]

File: client/components/AddNicknameForm.js [hunk 12]
```diff
@@ -102,7 +122,11 @@ class AddNicknameForm extends Component {
     Redux Properties
   \***************************************************************************/
 
-  static mapDispatchToProps = ['addNickname']
+  static mapDispatchToProps = ['addNickname', 'addNicknamePass']
```
Comment (UncleClapton) [inline client/components/AddNicknameForm.js]: When `mapDispatchToProps` is given an array of strings, it's telling the `@connect` decorator to map the function of the same name found within the files of the `/client/store/actions` directory. I feel like this whole 'lets define a magic array of strings' thing has been confusing for a lot of people who has seen our website. might just remove it entirely soon once we move to the new namespaced functions thing I keep squeaking about.

Long story short, 'addNicknamePass' is not a valid store function, and there's no reason for it to be there.

File: client/components/UserDecalPanel.js [hunk 13]
```diff
@@ -16,22 +16,21 @@ import Component from './Component'
...
 @connect
 class UserDetailsPanel extends Component {
   /***************************************************************************\
     Class Properties
   \***************************************************************************/
 
   state = {
-    checkingEligibility: true,
-    redeeming: false,
+    ...initialState,
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: No reason for an `initialState` variable to be defined here as we don't need to clear state for any reason.

File: client/components/UserDecalPanel.js [hunk 14]
```diff
@@ -87,25 +109,31 @@ class UserDetailsPanel extends Component {
...
+      <div className="code">
+        <button
+          aria-label="Show decal code"
+          className="icon toggle"
+          name={decal.id}
+          onClick={this._handleToggle}
+          type="button">
+          <FontAwesomeIcon icon={this.state.decalsVisible[decal.id] ? 'eye' : 'eye-slash'} fixedWidth />
+        </button>
+        {this.state.decalsVisible[decal.id] ? decal.attributes.code : `•••••-•••••-•••••-•••••-${decal.attributes.code.substring(24)}`}
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Could we multiline long ternary expressions like this?

```suggestion
        {
          this.state.decalsVisible[decal.id] 
            ? decal.attributes.code 
            : `•••••-•••••-•••••-•••••-${decal.attributes.code.substring(24)}`
        }
```

File: client/components/UserDecalPanel.js [hunk 15]
```diff
@@ -54,13 +53,36 @@ class UserDetailsPanel extends Component {
     })
   }
 
-  static _renderClaimedAtRow (row) {
-    const rescue = row.original
+  _handleToggle = (event) => {
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: I'd name this something more descriptive like
```suggestion
  _handleDecalVisibilityToggle = (event) => {
```

File: client/components/AddNicknameForm.js [hunk 16]
```diff
@@ -36,24 +39,28 @@ class AddNicknameForm extends Component {
...
+    const {
+      nickname,
+      password,
+    } = this.state
 
     event.preventDefault()
 
     this.setState({ submitting: true })
 
     /* eslint-disable no-alert */
```
Comment (UncleClapton) [inline client/components/AddNicknameForm.js]: we can remove this eslint disable and enable lines now that we are no longer using a prompt to collect the password.

File: client/components/ProfileHeader.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+import {
+  selectUser,
+  selectUserDisplayRat,
+  selectUserAvatar,
+} from '../store/selectors'
+
+
+
+
+// Component Constants
```
Comment (UncleClapton) [inline client/components/ProfileHeader.js]: Could you do a round of clean up on all the files changed/introduced by this PR and remove empty section comments like this?

File: client/components/ProfileHeader.js [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        </div>
+      </div>
+    )
+  }
+
+
+
+
+
+  /***************************************************************************\
```
Comment (UncleClapton) [inline client/components/ProfileHeader.js:88]: empty sections should not have headers like this.

File: client/components/ProfileHeader.js [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    Redux Properties
+  \***************************************************************************/
+
+  static mapDispatchToProps = ['getUser']
```
Comment (UncleClapton) [inline client/components/ProfileHeader.js]: getUser is unused and not needed here. Further, if the user is already authenticated there is no need to ever call this from a component.

Could just entirely remove this line

File: client/components/ProfileHeader.js [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+  state = {}
+
+
+
+
+
+  /***************************************************************************\
+    Private Methods
+  \***************************************************************************/
+  _renderUserGroups = () => this.props.user.relationships.groups.data.map((item) => <li className={`badge ${item.id}`} key={item.id}>{item.id}</li>)
```
Comment (UncleClapton) [inline client/components/ProfileHeader.js]: Please split this up a little. At least where it becomes JSX so it's not all forced onto one line.
Kind of like:
```suggestion
  _renderUserGroups = () => this.props.user.relationships.groups.data.map((item) => (
    <li className={`badge ${item.id}`} key={item.id}>
      {item.id}
    </li>
  ))
```

Though that might not be enough. If the user has no groups, data will most likely be null. You should probably run a null check before mapping.

File: client/components/UserDecalPanel.js [hunk 21]
```diff
@@ -116,18 +144,16 @@ class UserDetailsPanel extends Component {
...
-        loadingText={checkingEligibility ? 'Checking decal eligibility...' : 'Redeeming decal codes...'}
-        manual
-        minRows={5}
-        noDataText={this._renderNoDataText()}
-        showPagination={false} />
+      <div className="panel user-decals">
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this.renderDecalCodes(decals))}
+          {(decals.length) ? ('') : (this.renderNoDataText())}
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: I would personally write this like:
```suggestion
          {!decals.length && this.renderNoDataText()}
```

React won't render anything if a block evaluates to `false`, `null`, or `undefined`.

File: client/components/UserDecalPanel.js [hunk 22]
```diff
@@ -116,18 +144,16 @@ class UserDetailsPanel extends Component {
...
-        loading={checkingEligibility || redeeming}
-        loadingText={checkingEligibility ? 'Checking decal eligibility...' : 'Redeeming decal codes...'}
-        manual
-        minRows={5}
-        noDataText={this._renderNoDataText()}
-        showPagination={false} />
+      <div className="panel user-decals">
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this.renderDecalCodes(decals))}
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Another long ternary to be multi-lined

File: client/components/UserDecalPanel.js [hunk 23]
```diff
@@ -139,31 +165,6 @@ class UserDetailsPanel extends Component {
     Getters
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: Another empty getter section

File: client/components/UserNicknamesPanel.js [hunk 24]
```diff
@@ -4,55 +4,87 @@ import React from 'react'
...
-
+// Component constants
+const MAXNICKS = 20 // Maximum IRC Nicknames allowed
 
 
 
 @connect
 class UserNicknamesPanel extends Component {
-  _deleteNickname = async (event) => {
+  _handleDeleteNickname = async (event) => {
```
Comment (UncleClapton) [inline client/components/UserNicknamesPanel.js]: could we restructure this to follow the template, and add in the section headers?

Class Properties (like state)
Private Methods (_ methods/functions)
Public Methods
Getters
Redux Mapping Properties
Prop Definitions

File: client/components/UserNicknamesPanel.js [hunk 25]
```diff
@@ -4,55 +4,87 @@ import React from 'react'
...
     await this.props.deleteNickname(event.target.name)
     return 'Deleted!'
   }
 
+  state = {
+    formOpen: false,
+  }
+
+  _handleToggle = () => {
```
Comment (UncleClapton) [inline client/components/UserNicknamesPanel.js]: I'd rename this to something more descriptive like
```suggestion
  _handleFormVisibilityToggle = () => {
```

File: client/components/UserNicknamesPanel.js [hunk 26]
```diff
@@ -4,55 +4,87 @@ import React from 'react'
...
+      ...state,
+      formOpen: !state.formOpen,
+    }))
+  }
+
   render () {
     const {
       user,
     } = this.props
+    const maxNicksReached = (user.attributes.nicknames.length === MAXNICKS)
```
Comment (UncleClapton) [inline client/components/UserNicknamesPanel.js]: could we make this `>=` just for safety?

```suggestion
    const maxNicksReached = (user.attributes.nicknames.length >= MAXNICKS)
```

File: client/components/UserNicknamesPanel.js [hunk 27]
```diff
@@ -4,55 +4,87 @@ import React from 'react'
...
                   </ConfirmActionButton>
                   */}
                 </li>
               ))}
           </ul>
         </div>
-        <footer>
-          <AddNicknameForm />
-        </footer>
+        {/* (this.state.formOpen) && (
```
Comment (UncleClapton) [inline client/components/UserNicknamesPanel.js]: old garbage?

File: client/helpers/formatTime.js [hunk 28]
```diff
@@ -14,10 +14,13 @@ const formatAsEliteDateTime = (timestamp) => moment(timestamp).add(ELITE_GAME_YE
 
 const formatAsEliteDate = (timestamp) => moment(timestamp).add(ELITE_GAME_YEAR_DESPARITY, 'years').format('DD MMM YYYY').toUpperCase()
 
+const formatAsEliteDateLong = (timestamp) => moment(timestamp).add(ELITE_GAME_YEAR_DESPARITY, 'years').format('DD MMMM YYYY')
```
Comment (UncleClapton) [inline client/helpers/formatTime.js:17]: yum more date formatters! 😋

File: client/components/UserNicknamesPanel.js [hunk 29]
```diff
@@ -4,55 +4,87 @@ import React from 'react'
...
     return 'Deleted!'
   }
 
+  state = {
+    formOpen: false,
+  }
+
+  _handleToggle = () => {
+    this.setState((state) => ({
+      ...state,
```
Comment (UncleClapton) [inline client/components/UserNicknamesPanel.js]: no need to spread state like this normally. `setState` shallow merges the object you pass it. the only time we need to spread existing state is when we're trying to change something within an object stored within state, without losing the rest of that object.

This `setState` can just become
```javascript
this.setstate((state) => { formOpen: !state.formOpen })
```

File: client/components/ProfileHeader.js [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+    Public Methods
+  \***************************************************************************/
+
+  render () {
+    const {
+      displayRat,
+      userAvatar,
+    } = this.props
+
+    let { attributes } = this.props.user
```
Comment (UncleClapton) [inline client/components/ProfileHeader.js]: These 2 lines could actually be
```suggestion
    const attributes = this.props.user.attributes || {}
```

File: client/components/UserDecalPanel.js [hunk 31]
```diff
@@ -87,25 +109,31 @@ class UserDetailsPanel extends Component {
...
-    } = this.props
-
-    if (!eligible) {
-      await checkDecalEligibility()
+    if (!decals.length) {
+      return null
     }
 
-    this.setState({ checkingEligibility: false })
+    return decals.map(this.renderDecalCode)
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js]: you could actually write this bit out as 
```suggestion
    return Boolean(decals.length) && decals.map(this.renderDecalCode)
```
 and remove that `if (!decals.length)` block above.

File: client/helpers/formatTime.js [hunk 32]
```diff
@@ -14,10 +14,13 @@ const formatAsEliteDateTime = (timestamp) => moment(timestamp).add(ELITE_GAME_YE
...
+
 
 
 export {
   // eslint-disable-next-line import/prefer-default-export
```
Comment (UncleClapton) [inline client/helpers/formatTime.js]: Since we have more than one export from this file now, we can remove this config override.

**Review by UncleClapton — CHANGES_REQUESTED** [review]
Here comes the big one 🤣 

Overall really good stuff. I'm loving how this is turning out. As we discussed lets hold off on the last two items for another ticket. I'm going to remove them both and put them into separate tickets.

**Comment by kenneaal** [comment]
Don't mind me, I'm just here to test the announcer suppression.

**Comment by kenneaal** [comment]
> Don't mind me, I'm just here to test the announcer suppression.

Test 2.

**Comment by kenneaal** [comment]
Test 3

**Comment by kenneaal** [comment]
Test 4

**Comment by kenneaal** [comment]
And test 5...

**Comment by kenneaal** [comment]
Say hello to test 6!

**Comment by kenneaal** [comment]
Test 7 has joined the game.

**Comment by kenneaal** [comment]
Test 8 uses PleaseWork.

**Comment by kenneaal** [comment]
It's not very effective.

**Comment by kenneaal** [comment]
Test 9.

**Comment by kenneaal** [comment]
Test 10!

**Comment by kenneaal** [comment]
Why do you hate me, Python?

**Comment by itssimple** [comment]
This is a nice comment. 👌

**Comment by kenneaal** [comment]
Oh, more comments!

**Comment by kenneaal** [comment]
This should not appear.

**Comment by itssimple** [comment]
So fast I can't even!

**Comment by itssimple** [comment]
Fast enough for ya?

**Comment by kenneaal** [comment]
One comment, yes.

**Comment by kenneaal** [comment]
Two comment, yes yes.

File: client/components/UserDecalPanel.js [hunk 33]
```diff
@@ -16,28 +16,11 @@ import Component from './Component'
...
-const ELITE_GAME_YEAR_DESPARITY = 1286 // Years between IRL year and Elite universe year
 
 
 
-/***********************************************/
-/* DUE FOR REFACTORING. DO NOT MODIFY FURTHER. */
-/***********************************************/
 @connect
 class UserDetailsPanel extends Component {
-  /***************************************************************************\
```
Comment (UncleClapton) [inline client/components/UserDecalPanel.js:28]: We still need state, it just doesn't need to be in an initialState variable 🤣

**Review by UncleClapton — CHANGES_REQUESTED** [review]

**Comment by kenneaal** [comment]
These additional testing comments brought to you by Clapton.

**Comment by kenneaal** [comment]
Who is entirely to blame!

**Review by UncleClapton — APPROVED** [review]
LGTM!
